lib/lyber_core/base.rb

```diff
@@ -0,0 +1,76 @@
+require 'benchmark'
+require 'active_support'
+
+module LyberCore
+  # The skeleton of robots, a replacement for LyberCore::Robot
+  class Base
+    # Called by job-manager
+    # Instantiate the Robot and call #work with the passed in druid
+    # @param [String] druid
+    def self.perform(druid)
+      worker.work(druid)
```

Comment:
Maybe I'm not fully understanding things here, but given that implemented robots currently look like this:

```ruby
class MyRobot
    include LyberCore::Robot

   def initialize(opts = {})
        super(REPOSITORY, WORKFLOW_NAME, ROBOT_NAME, opts)
   end

   def perform(druid)
      robot_private_method_that_does_stuff
   end
end
```

would it not be simpler if this new class facilitated something like this?


```ruby
class MyRobot << LyberCore::BaseRobot
   def initialize(opts = {})
        super(REPOSITORY, WORKFLOW_NAME, ROBOT_NAME, opts)
   end

   def perform(druid)
      robot_private_method_that_does_stuff
   end
end
```

 I think we currently DO need to know the params for .new(), and I think we can continue with that burden, as it is crucial they are properly set to match the workflow definition.

So I *think* this implies something like this:

```ruby
# The skeleton of robots, a replacement for LyberCore::Robot
class Base
     # note that repo, workflow_name and step_name all need to match the workflow definition, and 
    # also need to match the declared classes of your robots (e.g. if your workflow is dor_awesomeWF,
    # then repo = 'dor' and workflow_name = 'awesome' and step_name matches the declared step name in
    # your workflow definition XML in that datastream for the workflow object.
    # note further that your robot must be Robots::DorRepo::Awesome::MyRobot where the first 3 are modules. 
    def initialize(repo, workflow_name, step_name, opts = {})
      # is it possible to insert code to raise NotImplementedError if it wasn't called from a subclass? 
      # code goes here
    end

      # Called by job-manager on derived-class
      # Instantiate the Robot and call #work with the passed in druid
      def perform(druid)
       # is it possible to insert code to raise NotImplementedError if it wasn't called from a subclass? 
        bot = new
        bot.work druid
      end

      # all else the same?  
      #  wondering if "work" method can be private?

end
```

Reply:
@ndushay: Are you asking "would it be simpler if a base class forces a subclass to declare multiple constants?"  If that is a question, my answer is no, that is a part of what I am trying to get away from.

The `initialize` signature you propose is exactly what I have at line 22, so I see no problem there.  No need to raise `NotImplementedError` (or another), because many robots will never need to go beyond the base implementation.

The use of a `worker` class method allows for clean API seam as used in the tests.  I specifically want to avoid the `bot = new` because in the existing Module, that is a bogus piece of code.  It can never succeed, except by forcing the including Class to override `initialize`.  Here it can never succeed, except by the same.  That is the thing to be avoided.

In general I think the review standard here should be "is this better than what we do now?"  I think it clearly is.